src/dmd/errors.d

```diff
@@ -35,13 +35,69 @@ enum Classification : Color
     tip = Color.brightGreen,          /// for tip messages
 }
 
+alias MessageFunc = extern (C++) void function(const ref Loc loc, const(char)* text);
+
+private struct FragmentInfo
+{
+    const(char)* start; /// points to the first character of the `FRAGMENT_PREFIX_START` prefix for this fragment.
+
+    const(char)* name;
+    size_t nameLen;
+
+    const(char)* params; /// can be null.
+    size_t paramsLen;
+
+    const(char)* innerText;
+    size_t innerTextLen;
+}
+
+private immutable FRAGMENT_PREFIX = "\033¬\r!\033¬\t*"; // super random series of weird characters that no user code would ever really use. Change to e.g. "__PREFIX__" when debugging.
+private immutable FRAGMENT_PREFIX_START = FRAGMENT_PREFIX ~ "$[";
+private immutable FRAGMENT_PREFIX_BODY_END = FRAGMENT_PREFIX ~ '}';
+
+/**
+ * Print an error message using the extended string format, increasing the global error count.
+ * Params:
+ *      loc    = location of error
+ *      format = extended format specification
+ *      ...    = printf-style variadic arguments
+ */
+static if (__VERSION__ < 2092)
+    extern (C++) void errorEx(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list ap;
+        va_start(ap, format);
+        verrorEx(loc, format, ap);
+        va_end(ap);
+    }
+else
+    pragma(printf) extern (C++) void errorEx(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list ap;
+        va_start(ap, format);
+        verrorEx(loc, format, ap);
+        va_end(ap);
+    }
+
+/**
+ * Similar to `verror` except it uses the extended format string described by `verrorFormatPrint`.
+ * Params:
+ *      loc    = location of error
+ *      format = extended format specification
+ *      ap     = printf-style argument list
+ */
+extern (C++) void verrorEx(const ref Loc loc, const(char)* format, va_list ap)
+{
+    verrorFormatPrint(loc, format, ap, &_verrorExPrint, &_verrorExPrintSupplemental);
+}
+extern (C++) private void _verrorExPrint(const ref Loc loc, const(char)* text) { error(loc, "%s", text); }
+extern (C++) private void _verrorExPrintSupplemental(const ref Loc loc, const(char)* text){ errorSupplemental(loc, "%s", text); }
```

Comment:
Also these two functions.  If you are just forwarding to extern(C++) functions now, then I don't see any need to keep these.

Reply:
Ah, good point. I think I was just blindly copying the style of the other error funcs.